### PR TITLE
Allow configurable auth0

### PR DESCRIPTION
**Pull request by lorentzlasson**
---

Tested on devices 
- [ ] Desktop 💻
- [ ] Mobile 📱

Tests  
- [ ] All tests are running ✔️
- [ ] Test are updated 🧪
- [ ] Code Review 👩‍💻
- [ ] QA 👌

⏲️ Time spent on CR:

⏲️ Time spent on QA:

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -2,3 +2,9 @@ DB_NAME='EffektDonasjonDB_Local'
 DB_USER='root'
 DB_PASS='effekt'
 ALLOWED_PRODUCTION_ORIGINS='http://localhost:3000'
+
+# Keep in sync with values configured in Auth0
+AUTH_AUDIENCE=https://data.gieffektivt.no
+AUTH_BASE_URL=https://gieffektivt.eu.auth0.com/
+AUTH_ROLE_CLAIM=https://gieffektivt.no/roles
+AUTH_USER_ID_CLAIM=https://gieffektivt.no/user-id
```
Comment (lorentzlasson) [inline .env.example:10]: Keeping NO as example :shrug:

File: cloud.dev.deploy.yaml [hunk 2]
```diff
@@ -1,14 +1,44 @@
...
+      - 'deploy'
+      - 'geeffektivt-dev'
+      - '--image'
+      - 'gcr.io/$PROJECT_ID/effektapi-dev:latest'
+      - '--region'
+      - 'europe-north1'
+      - '--platform'
+      - 'managed'
+      - '--allow-unauthenticated'
+
```
Comment (lorentzlasson) [inline cloud.dev.deploy.yaml:43]: Example run: https://console.cloud.google.com/cloud-build/builds;region=global/7cf8fce0-c6ea-4ecb-8b82-f9e7d85ff372;tab=detail?authuser=1&project=hidden-display-243419

(NO failing here for some reason though)

File: src/custom_modules/authorization/authMiddleware.ts [hunk 3]
```diff
@@ -33,11 +35,21 @@ export const auth = (permission, api = true) => {
...
 
-export const checkDonor = (donorId, req, res, next) => {
-  const handler = claimCheck(
-    (claims) =>
-      (claims as ExtendedJWTPayload)[roleClaim].includes("admin") ||
-      claims[useridClaim] === donorId
+const userIsAdmin = (claims: JWTPayload) => {
+  const roleClaims = claims[roleClaim];
+  if (!Array.isArray(roleClaims)) return false;
+  return roleClaims[roleClaim].includes("admin");
```
Comment (lorentzlasson) [inline src/custom_modules/authorization/authMiddleware.ts:41]: Can't have the `ExtendedJWTPayload` anymore as the keys are now dynamic, but the type safety of this should be higher even as it doesn't rely on type assertions

File: .env.example (hunk 1 above)
Comment (lorentzlasson) [reply .env.example:10]: Will need to add these on all cloud runs before deploying

File: src/custom_modules/authorization/authMiddleware.ts [hunk 4]
```diff
@@ -6,21 +6,23 @@ import {
...
+import {
+  authAudience,
+  authRoleClaim,
+  authIssuerBaseURL,
+  authUserIdClaim,
+} from "../../config";
 
 const authorizationRoles = require("../../enums/authorizationRoles.js");
 
+// Defaulting to "default" to enable tests to run
```
Comment (lorentzlasson) [inline src/custom_modules/authorization/authMiddleware.ts:18]: The test suite doesn't seem to case at all what values are here, but the `auth0` throws if the values are empty.

Doesn't feel ideal, but it works. Any better ideas?

**Review by adamalfredsson — APPROVED** [review]
Looks nice and clear to me! Not confident in reviewing the deploy configuration again, though!

File: cloud.dev.deploy.yaml (hunk 2 above)
Comment (lorentzlasson) [reply cloud.dev.deploy.yaml:43]: I don't think it affects stability, but allows the deploys to happen in parallel

File: src/custom_modules/authorization/authMiddleware.ts [hunk 5]
```diff
@@ -33,11 +35,21 @@ export const auth = (permission, api = true) => {
...
+
+const userIsDonor = (claims: JWTPayload, donorId: number) =>
+  claims[userIdClaim] === donorId;
+
+const userIsAllowedToManageDonor = (claims: JWTPayload, donorId: number) =>
+  userIsAdmin(claims) || userIsDonor(claims, donorId);
+
+export const checkDonor = (donorId: number, req, res, next) => {
+  const handler = claimCheck((claims) =>
+    userIsAllowedToManageDonor(claims, donorId)
```
[the remark this answers is not in the document]
Comment (lorentzlasson) [reply src/custom_modules/authorization/authMiddleware.ts]: Good observation. Perhaps definite form would help (`userIsTheDonor`)? :thinking:

File: cloud.dev.deploy.yaml (hunk 2 above)
Comment (lorentzlasson) [reply cloud.dev.deploy.yaml:43]: The failing :norway: deploy has been fixed. Turned out to be unrelated

**Review by fellmirr — APPROVED** [review]
Great job!

**Pull request merged**